keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,972 @@
+and make progress.
+-->
+
+* Address more specific issues with any particular pod status field(s). 
+* Expand the set of mutable fields.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### API Changes
+
+#### Generation and ObservedGeneration
+
+The `metadata.generation` field is currently unused on pods, but we can start 
+leveraging it to help track which pod state is currently being reflected in the 
+pod status. [`metadata.generation`](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata) is documented as "a sequence
+number representing a specific generation of the desired state. Set by the 
+system and monotonically increasing, per-resource." For consistency, pod 
+`metadata.generation` will be incremented whenever the pod has changes that the kubelet
+needs to actuate. 
+
+A new optional field `status.observedGeneration` field will be added to the pod status.
+Kubelet will set this to communicate which pod state is being expressed in the 
+current pod status. This is analogous to the `status.observedGeneration` that exists
+in other resources' statuses such as [StatefulSets](https://github.com/kubernetes/api/blob/9e7d345b161c12a3056efa88e285b3ef68450c54/apps/v1/types.go#L274).
+
+Likewise, a new optional `observedGeneration` field will be added to the pod's 
+`status.condition` struct. This is to keep parity with the 
+[metav1.Condition struct](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1620-L1625). 
+
+The net result will be a new `status.observedGeneration` field for the kubelet
+to express which generation the top-level status relates to, and a new `status.conditions[i].observedGeneration`
+field for the writer of that condition to express which generation that condition
+relates to.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Because there is only one singular `status.observedGeneration` in the pod status, only
+one writer can set it. This is consistent with other object types that have
+`status.observedGeneration`, and the expectation is that the primary controller for
+the object sets the field. For a pod, that primary controller is the kubelet, so
+once a pod is bound to a node, we expect that the kubelet on that node is the
+sole writer of `status.observedGeneration`. 
```

Comment:
So when node lifecycle controller sets the `NotReady` for the Pod that is associated with the disconnected Node, it will not set observedGeneration?

Reply:
That's a condition, right? It should probably set the `observedGeneration` on the condition.

Reply:
Yeah, if it's a condition, writers of conditions should set `observedGeneration` on the condition. For the scope of this KEP, only the kubelet will write to the top-level `status.observedGeneration`. 

We briefly talked about having the scheduler set `observedGeneration` before the pod is bound to a node, but I'm not sure if that should be part of this KEP or if that should be a future conversation.

Reply:
> We briefly talked about having the scheduler set `observedGeneration` before the pod is bound to a node, but I'm not sure if that should be part of this KEP or if that should be a future conversation.

I think the scheduler *primarily* sets conditions, so sweeping it to populate condition.observedGeneration as part of the general follow-up to sweep pod condition writers would make sense to me

Reply:
> This also can make it such that kubelet will run behind in the observedGenerations when kubelet reconnects. Will the status update from older generation be ignored?

Not sure I follow; do you mind elaborating on the scenario here?

Reply:
I think this is about the kubelet restart scenario. Since kubelet doesn't checkpoint the full pod specs, it'd still wait to connect to the apiserver first, and it shouldn't be sending statuses without accessing the pod specs, but correct me if I'm wrong.

Reply:
scheduler and node lifecycle controllers may set conditions or status. So those will need to update the observedGeneration as well. Those are well-known controllers of Pods alongside the kubelet

Reply:
> scheduler and node lifecycle controllers may set conditions or status. 

Based on this and [along with the suggestion below to have populating `condition.observedGeneration` in-scope for this KEP](https://github.com/kubernetes/enhancements/pull/5068#discussion_r1947127406), there are a few things we need to add for scheduler and node lifecycle controller:
- They will also need to have the `PodObservedGenerationTracking` feature flag
- When scheduler or node lifecycle controller populate a pod _condition_, they need to set the `observedGeneration` on that condition
- When scheduler or node lifecycle controller update the pod status beyond just updating a pod condition, they will need to set the top-level `status.observedGeneration`

I added this to the body of the KEP as well.

Reply:
thanks!